README.md
# Policy Query Script and Vulnerability Reporting

## Deployment Objects

The following resources will be deployed to the resource group as defined in the script:

- Automation account (managed identity enabled)
- PowerShell modules for the Automation Account - Az.Accounts, Az.Resources, Az.ResourceGraph, Az.Storage
- Automation account variables
- Storage account to hold CSV files
- Containers in the storage account to hold the runbooks, scans, output CSV files
- 2 runbooks to the Automation account
- A Logic App to send reports
- Connection to Azure Storage and Microsoft 365 for the Logic App

## Deployment instructions

1) Ensure the [Bicep](https://github.com/Azure/bicep) is installed and available at $Path
2) Deploy the solution by modifying the parameter value when deploying - an example is below:

```
.\deploy.ps1 -ManagementGroupName eslz `
        -DeploymentResourceGroupName eslz-auto `
        -AutomationAccountName eslz-aauto `
        -StorageAccountName eslzstor01auto `
        -AutomationAccountLocation australiaeast `
        -StorageAccountLocation australiasoutheast
```

## Post Deployment

1) M365 Logic App connection will need to be authorised
- For the M365 connection you need to provide authentication
2) Provide an email address in the Logic App activity that reports will be sent to
3) Adjust the frequency for the Logic App to run - i.e. if you are only generating a report weekly, set it to run daily - if no new blobs are available it just skips the trigger
4) Create [Automation Schedules](https://docs.microsoft.com/en-us/azure/automation/shared-resources/schedules) and link to each runbook
5) Give the managed identity for the automation account - permission to read policies and access the resource graph at the management group level. Access for the identity on the storage account should be granted by the script

**NOTE:** The Logic app only looks at one container in the storage account - you may have to duplicate it and point the second instance to a different container
if you are keeping the scans and policy compliance CSV files separately. 



CHANGELOG.md
v0.2
- Secured storage account, access is granted during runbook execution only to the runbook worker
- Logic app now has managed identity configured, firewall access to storage account is granted

deploy.ps1
<#
.SYNOPSIS
    Deploys the solution as at https://github.com/anwather/policyQueryScript
.DESCRIPTION
    See README.md for more details
.EXAMPLE
    PS C:\> deploy.ps1 -ManagementGroupName eslz `
        -DeploymentResourceGroupName eslz-auto `
        -AutomationAccountName eslz-aauto `
        -StorageAccountName eslzstor01auto `
        -AutomationAccountLocation australiaeast `
        -StorageAccountLocation australiasoutheast
.PARAMETER ManagementGroupName
    This is the root management group used to query the Resource Graph and query policy definitions
.PARAMETER DeploymentResourceGRoupName
    Resource group for the resources to be deployed into
.PARAMETER AutomationAccountName
    New automation account to be created
.PARAMETER StorageAccountName
    New storage account to be created
.PARAMETER AutomationAccountLocation
    Location for automation account
.PARAMETER StorageAccountLocation
    Location for storage account
    The storage account and automation account should not be in the same region
#>

Param(
    [Parameter(Mandatory = $true)]
    [string]$ManagementGroupName,
    [Parameter(Mandatory = $true)]
    [string]$DeploymentResourceGroupName,
    [Parameter(Mandatory = $true)]
    [string]$AutomationAccountName,
    [Parameter(Mandatory = $true)]
    [string]$StorageAccountName,
    [Parameter(Mandatory = $true)]
    [string]$AutomationAccountLocation,
    [Parameter(Mandatory = $true)]
    [string]$StorageAccountLocation
)

if ($AutomationAccountLocation -eq $storageAccountLocation) {
    Write-Error "Storage account and automation account cannot be in the same location"
    exit
}

try {
    bicep --version
}
catch {
    Write-Error "Ensure Bicep is available 'https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/install#install-manually'"
    exit
}



New-AzResourceGroupDeployment -ResourceGroupName $deploymentResourceGroupName `
    -TemplateFile .\azuredeploy.bicep `
    -StorageAccountName $storageAccountName `
    -AutomationAccountName $automationAccountName `
    -ManagementGroupName $managementGroupName `
    -StorageAccountResourceGroupName $deploymentResourceGroupName `
    -StorageAccountLocation $storageAccountLocation `
    -AutomationAccountLocation $automationAccountLocation `
    -Verbose

$ctx = (Get-AzStorageAccount -ResourceGroupName $deploymentResourceGroupName -StorageAccountName $storageAccountName).Context

$runbooks = @("queryPolicyCompliance", "queryVulnerabilities")

$runbooks | ForEach-Object {
    Set-AzStorageBlobContent -File ".\$_.ps1" -Blob "$_.ps1" -Container runbooks -Context $ctx -Force

    $token = New-AzStorageBlobSASToken -Blob "$_.ps1" -Container runbooks -Permission rl -ExpiryTime (Get-Date).AddMinutes(5) -Context $ctx -FullUri

    New-AzResourceGroupDeployment -ResourceGroupName $deploymentResourceGroupName `
        -TemplateFile .\runbookdeploy.bicep `
        -RunbookName $_ `
        -RunbookUri $token `
        -AutomationAccountName $automationAccountName `
        -AutomationAccountLocation $automationAccountLocation `
        -Verbose
}

Update-AzStorageAccountNetworkRuleSet -ResourceGroupName $deploymentResourceGroupName -Name $storageAccountName -DefaultAction Deny

$deployment = New-AzResourceGroupDeployment -ResourceGroupName $deploymentResourceGroupName `
    -TemplateFile .\logicApp.bicep `
    -Verbose `
    -StorageAccountName $storageAccountName

$resourceAccessRule = @{
    TenantId   = (Get-AzContext).Tenant
    ResourceId = $deployment.outputs.logicAppResourceId.Value
}

Update-AzStorageAccountNetworkRuleSet -ResourceGroupName $deploymentResourceGroupName -Name $storageAccountName -ResourceAccessRule $resourceAccessRule



queryPolicyCompliance.ps1
Connect-AzAccount -Identity

$complianceQuery = @"
policyresources
| where type == 'microsoft.policyinsights/policystates'
| where properties.complianceState == 'NonCompliant'
| join kind = inner (resourcecontainers
| where type == "microsoft.resources/subscriptions"
| project subscriptionId,name) on subscriptionId
| project-away id, name, type, tenantId, kind, location, resourceGroup, managedBy, sku, plan, tags, identity, zones, extendedLocation, subscriptionId1, apiVersion
| project-rename subscriptionName=name1
"@

$managementGroup = Get-AutomationVariable "ManagementGroupName"
$storageAccountName = Get-AutomationVariable "StorageAccountName"
$storageAccountResourceGroupName = Get-AutomationVariable "StorageAccountResourceGroupName"

$objects = @()

do {
    if ($null -eq $SkipToken) {
        $results = Search-AzGraph -Query $complianceQuery -First 1000 -ManagementGroup $managementGroup
        foreach ($res in $results) {
            $objects += $res
        }
        if ($results.SkipToken) {
            $SkipToken = $results.SkipToken
        }
    }
    else {
        $results = Search-AzGraph -Query $complianceQuery -First 1000 -SkipToken $SkipToken -ManagementGroup $managementGroup
        foreach ($res in $results) {
            $objects += $res
        }
        if ($results.SkipToken) {
            $SkipToken = $results.SkipToken
        }
        else {
            $SkipToken = $null
        }
    }
}
until ($null -eq $SkipToken)

$definitions = Get-AzPolicyDefinition

$polMap = @{}

$definitions | Foreach-Object {
    $polMap.Add($_.ResourceId, @{
            Name        = $_.Properties.DisplayName
            Description = $_.Properties.Description
        })
}

$outputObjects = @()

$objects | Foreach-Object {
    $o = [PSCustomObject]@{
        SubscriptionId    = $_.Properties.SubscriptionId
        SubscriptionName  = $_.SubscriptionName
        ResourceType      = $_.Properties.ResourceType 
        ResourceGroup     = $_.Properties.ResourceGroup
        ResourceId        = $_.Properties.ResourceId
        PolicyName        = $polMap[$_.Properties.PolicyDefinitionId].Name
        PolicyDescription = $polMap[$_.Properties.PolicyDefinitionId].Description
    }
    $outputObjects += $o
}

Write-Output "Reporting $($outputObjects.Count) non compliant objects"

$outputObjects | Export-CSV -Path "$env:Temp\output.csv" -NoTypeInformation

$runbookWorkerIp = (Invoke-WebRequest -Uri "https://api.ipify.org" -UseBasicParsing).Content

Add-AzStorageAccountNetworkRule -ResourceGroupName $storageAccountResourceGroupName -AccountName $storageAccountName -IPAddressOrRange $runbookWorkerIp

Start-Sleep -Seconds 60

$ctx = (Get-AzStorageAccount -ResourceGroupName $storageAccountResourceGroupName -StorageAccountName $storageAccountName).Context

$fileName = "$(Get-Date -Format yyyy_MM_dd_mm)_output.csv"

Set-AzStorageBlobContent -File "$env:Temp\output.csv" -Blob $fileName -Container output -Context $ctx -Force

Remove-AzStorageAccountNetworkRule -ResourceGroupName $storageAccountResourceGroupName -AccountName $storageAccountName -IPAddressOrRange $runbookWorkerIp





queryVulnerabilities.ps1
Connect-AzAccount -Identity

function ConvertPSObjectToHashtable {
    param (
        [Parameter(ValueFromPipeline)]
        $InputObject
    )

    process {
        if ($null -eq $InputObject) { return $null }

        if ($InputObject -is [System.Collections.IEnumerable] -and $InputObject -isnot [string]) {
            $collection = @(
                foreach ($object in $InputObject) { ConvertPSObjectToHashtable $object }
            )

            Write-Output -NoEnumerate $collection
        }
        elseif ($InputObject -is [psobject]) {
            $hash = @{}

            foreach ($property in $InputObject.PSObject.Properties) {
                $hash[$property.Name] = ConvertPSObjectToHashtable $property.Value
            }

            $hash
        }
        else {
            $InputObject
        }
    }
}

$query = @"
securityresources
 | where type == "microsoft.security/assessments"
 | where * contains "vulnerabilities in your virtual machines"
 | summarize by assessmentKey=name //the ID of the assessment
 | join kind=inner (
    securityresources
     | where type == "microsoft.security/assessments/subassessments"
     | extend assessmentKey = extract(".*assessments/(.+?)/.*",1,  id)
 ) on assessmentKey
| join kind = inner (resourcecontainers
| where type == "microsoft.resources/subscriptions"
| project subscriptionId,name) on subscriptionId
| project assessmentKey, subassessmentKey=name, id, parse_json(properties), resourceGroup, subscriptionId, tenantId, subscriptionName=name1
| extend description = properties.description,
         displayName = properties.displayName,
         resourceId = properties.resourceDetails.id,
         resourceSource = properties.resourceDetails.source,
         category = properties.category,
         severity = properties.status.severity,
         code = properties.status.code,
         timeGenerated = properties.timeGenerated,
         remediation = properties.remediation,
         impact = properties.impact,
         vulnId = properties.id,
         additionalData = properties.additionalData
| project-away properties, assessmentKey, subassessmentKey, id, tenantId
"@

$managementGroup = Get-AutomationVariable "ManagementGroupName"
$storageAccountName = Get-AutomationVariable "StorageAccountName"
$storageAccountResourceGroupName = Get-AutomationVariable "StorageAccountResourceGroupName"

$ctx = (Get-AzStorageAccount -ResourceGroupName $storageAccountResourceGroupName -StorageAccountName $storageAccountName).Context

$objects = @()

do {
    if ($null -eq $SkipToken) {
        $results = Search-AzGraph -Query $query -First 1000 -ManagementGroup $managementGroup
        foreach ($res in $results) {
            $objects += $res
        }
        if ($results.SkipToken) {
            $SkipToken = $results.SkipToken
        }
    }
    else {
        $results = Search-AzGraph -Query $query -First 1000 -SkipToken $SkipToken -ManagementGroup $managementGroup
        foreach ($res in $results) {
            $objects += $res
        }
        if ($results.SkipToken) {
            $SkipToken = $results.SkipToken
        }
        else {
            $SkipToken = $null
        }
    }
}
until ($null -eq $SkipToken)

$computerObjects = $objects | Group-Object -Property resourceId | Select-Object -ExpandProperty Name

Write-Output "Discovered $($computerObjects.Count) assessments"

$runbookWorkerIp = (Invoke-WebRequest -Uri "https://api.ipify.org" -UseBasicParsing).Content

Add-AzStorageAccountNetworkRule -ResourceGroupName $storageAccountResourceGroupName -AccountName $storageAccountName -IPAddressOrRange $runbookWorkerIp

Start-Sleep -Seconds 60

$computerObjects | Foreach-Object {
    $outputObjects = @()
    $objects | Where-Object resourceId -eq $_ | Foreach-Object {
        $o = [PSCustomObject]@{}
        $hashObject = $_ | ConvertPSObjectToHashtable
        $hashObject.GetEnumerator() | Foreach-Object {
            $o | Add-Member -MemberType NoteProperty -Name $_.Key -Value $_.Value
        }
        $outputObjects += $o
    }
    $fileName = "$($_.Split("/")[-1])_$($_.Split("/")[2])_$($_.Split("/")[4].ToLower())_$(Get-Date -Format yyyy_MM_dd_mm)_output.csv"
    $outputObjects | Export-CSV -Path $env:Temp\$fileName -NoTypeInformation
    Set-AzStorageBlobContent -File "$env:Temp\$fileName" -Blob $fileName -Container scans -Context $ctx -Force
}

Remove-AzStorageAccountNetworkRule -ResourceGroupName $storageAccountResourceGroupName -AccountName $storageAccountName -IPAddressOrRange $runbookWorkerIp